pkg/controller/node/controller_utils.go

```diff
@@ -243,6 +253,35 @@ func markAllPodsNotReady(kubeClient clientset.Interface, nodeName string) error
 	return fmt.Errorf("%v", strings.Join(errMsg, "; "))
 }
 
+//isPodForgivable will check whether Pod need to be forgive. Return true if pod has forgiveness toleration and duration is greater than forgivenessSeconds
+func isPodForgivable(pod api.Pod, nodeReadyTransitionTimestamp unversioned.Time) bool {
+	tolerations, err := api.GetTolerationsFromPodAnnotations(pod.ObjectMeta.Annotations)
+	if err != nil {
+		return false
+	}
+	var forgivenessToleration api.Toleration
+	found := false
+	for _, toleration := range tolerations {
+		if toleration.Key == unversioned.TaintNodeDown && toleration.Effect == api.TaintEffectNoExecute {
```

Comment:
what if pod was removed not because of TaintNodeDown? I should be able to add a taint with effect NoExecute without any problem, and it will have no effect because of this condition

I think you need to compare node taints with pods tolerations and use forgiveness only if certain taint is present on a node